r/R/util.R

```diff
@@ -110,3 +110,11 @@ handle_embedded_nul_error <- function(e) {
   }
   stop(e)
 }
+
+is_writable_table <- function(x) {
+  inherits(x, c("data.frame", "ArrowTabular"))
+}
+
+attr(is_writable_table, "fail") <- function(call, env){
+  paste0(eval(substitute(substitute(y, env), list(y = call$x)))," is not an object of class 'data.frame', 'RecordBatch', or 'Table'.")
```

Comment:
What do you think about saying "must be" instead of "is not"?

Also, previously you had said what the class of `x` was: is that feasible here? (Ok if not or not worth it, just observing that it was nice.)

Reply:
Also, I don't think the eval(substitute(substitute())) is totally safe: 

```
> assert_that(is_writable_table(c("a", "b")))
Error in stop(assertError(attr(res, "msg"))) : bad error message
```

`deparse(call$x)` seems to work:

```
> attr(is_writable_table, "fail") <- function(call, env){
  paste0(deparse(call$x)," is not an object of class 'data.frame', 'RecordBatch', or 'Table'.")
}
> assert_that(is_writable_table(c("a", "b")))
Error: c("a", "b") is not an object of class 'data.frame', 'RecordBatch', or 'Table'.
> assert_that(is_writable_table(letters))
Error: letters is not an object of class 'data.frame', 'RecordBatch', or 'Table'.
```

Reply:
I think that "must be" sounds better than "is not" after reading the tidyverse style guide's approach to errors since updating this PR ;) Yeah, makes sense - is a little more informative, will update.

I'll have a look into alternatives.  My problem was that when I try it with `deparse(call$x)` it works when called directly like you have above, but when I run the unit tests and it's being called from the `write_*` functions, the error message then reads `x must be an object of class 'data.frame', 'RecordBatch', or 'Table'.`.  I guess it's looking at the wrong level, but I wasn't sure how to get it to look at the right one.  Any ideas?

Reply:
I don't think that's bad because `x` is the name of the argument in these functions. If the argument were named something else, it would use that name:

```
> f <- function(zzz) {
  assert_that(is_writable_table(zzz))
}
> f(letters)
Error: zzz is not an object of class 'data.frame', 'RecordBatch', or 'Table'.
> f(c("a", "b"))
Error: zzz is not an object of class 'data.frame', 'RecordBatch', or 'Table'.
```

This is how `assertthat` does it too. See https://github.com/hadley/assertthat/blob/master/R/base-is.r for example.

Reply:
Ah, thanks, had got mixed up with something else you said, but makes sense now.

Worked out a way of putting the object's class back into the error message so have pushed that change now, as well as the comment about what that attribute does.